jdk/src/share/classes/sun/security/ec/ECPrivateKeyImpl.java

```diff
@@ -213,4 +226,55 @@ protected void parseKeyBits() throws InvalidKeyException {
             throw new InvalidKeyException("Invalid EC private key", e);
         }
     }
+
+    /**
+     * Returns the native EC public key context pointer.
+     * @return the native EC public key context pointer or -1 on error
+     */
+    long getNativePtr() {
+        if (nativeECKey == 0x0) {
+            synchronized (this) {
+                if (nativeECKey == 0x0) {
+                    ECParameterSpec parameters = this.getParams();
+                    ECPoint generator = parameters.getGenerator();
+                    EllipticCurve curve = parameters.getCurve();
+                    ECField field = curve.getField();
+                    byte[] a = curve.getA().toByteArray();
+                    byte[] b = curve.getB().toByteArray();
+                    byte[] gx = generator.getAffineX().toByteArray();
+                    byte[] gy = generator.getAffineY().toByteArray();
+                    byte[] n = parameters.getOrder().toByteArray();
+                    byte[] h = BigInteger.valueOf(parameters.getCofactor()).toByteArray();
+                    byte[] p = new byte[0];
+                    if (field instanceof ECFieldFp) {
+                        p = ((ECFieldFp)field).getP().toByteArray();
+                        nativeECKey = nativeCrypto.ECEncodeGFp(a, a.length, b, b.length, p, p.length, gx, gx.length, gy, gy.length, n, n.length, h, h.length);
+                    } else if (field instanceof ECFieldF2m) {
+                        p = ((ECFieldF2m)field).getReductionPolynomial().toByteArray();
+                        nativeECKey = nativeCrypto.ECEncodeGF2m(a, a.length, b, b.length, p, p.length, gx, gx.length, gy, gy.length, n, n.length, h, h.length);
+                    }
+                    if (nativeECKey > 0)  {
+                        Cleaner.create(this, new ECCleanerRunnable(nativeECKey));
+                        byte[] value = this.getS().toByteArray();
+                        if (nativeCrypto.ECCreatePrivateKey(nativeECKey, value, value.length) < 0) {
+                            nativeECKey = -1;
+                        }
+                    }
+                }
+            }
+        }
+        return nativeECKey;
+    }
+
+    static final class ECCleanerRunnable implements Runnable {
+        private final long ECKey;
+
+        public ECCleanerRunnable(long key) {
+            this.ECKey = key;
+        }
+
+        public void run() {
+            nativeCrypto.ECDestroyKey(ECKey);
+        }
+    }
```

Comment:
Rather than repeat this, I suggest adding this to `NativeCrypto`:
```
    public void createECCleaner(Object owner, long key) {
        Cleaner.create(owner, new Runnable() {
            @Override
            public void run() {
                NativeCrypto.this.ECDestroyKey(key);
            }
        });
    }
```
Then statements like
```
                        Cleaner.create(this, new ECCleanerRunnable(nativeECKey));
```
can be replaced by
```
                        nativeCrypto.createECCleaner(this, nativeECKey);
```

Reply:
i recommend `createECKeyCleaner(...)`

Reply:
done, thank you